### PR TITLE
Annotate WebAssembly.Authentication

### LINKED ISSUE #30751: Annotate and mark remaining Blazor WASM assemblies
* [ ] Microsoft.Authentication.WebAssembly.Msal
* [ ] Microsoft.AspNetCore.Components.WebAssembly.Authentication

### LINKED ISSUE #35162: Blazor WebAssembly MSAL authentication error
### Describe the bug

I have a [Blazor WebAssembly application](https://github.com/IEvangelist/learning-blazor) that relies on:

`<PackageReference Include="Microsoft.Authentication.WebAssembly.Msal" Version="6.0.0-preview.6.21355.2" />`

For authentication, with an AAD B2C instance. It works perfectly, when running locally. But now that the app is deployed to Azure as a static web app, I get the following error:

> "There was an error trying to log you in: 'Cannot read property 'toLowerCase' of undefined'"

The website is [deployed here](https://purple-stone-0b2589410.azurestaticapps.net).

### To Reproduce

Deploy a Blazor WebAssembly app (with PWA bits) to Azure static web apps, with AAD B2C auth. Try logging in.

### Exceptions (if any)

None

### Further technical details

Using .NET Core SDK Version: 6.0.100-preview.6.21355.2

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/dotnet/aspnetcore/issues/30751
Fixes https://github.com/dotnet/aspnetcore/issues/35162

### CODE REVIEW

File: src/Components/WebAssembly/Authentication.Msal/src/MsalWebAssemblyServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -35,7 +35,8 @@ public static class MsalWebAssemblyServiceCollectionExtensions
     /// <param name="services">The <see cref="IServiceCollection"/>.</param>
     /// <param name="configure">A callback to configure the <see cref="RemoteAuthenticationOptions{MsalProviderOptions}"/>.</param>
     /// <returns>The <see cref="IServiceCollection"/>.</returns>
-    public static IRemoteAuthenticationBuilder<TRemoteAuthenticationState, RemoteUserAccount> AddMsalAuthentication<TRemoteAuthenticationState>(this IServiceCollection services, Action<RemoteAuthenticationOptions<MsalProviderOptions>> configure)
+    public static IRemoteAuthenticationBuilder<TRemoteAuthenticationState, RemoteUserAccount> AddMsalAuthentication<[DynamicallyAccessedMembers(JsonSerialized)] TRemoteAuthenticationState>(
```
Comment (TanayParikh) [inline src/Components/WebAssembly/Authentication.Msal/src/MsalWebAssemblyServiceCollectionExtensions.cs:38]: nit;
```suggestion
    public static IRemoteAuthenticationBuilder<TRemoteAuthenticationState, RemoteUserAccount> AddMsalAuthentication<
    [DynamicallyAccessedMembers(JsonSerialized)] TRemoteAuthenticationState>(
```

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj [hunk 2]
```diff
@@ -75,14 +76,7 @@
...
 
-    <DefineStaticWebAssets Condition="@(_InteropBuildOutput) != ''"
-      SourceType="Computed"
-      SourceId="$(PackageId)"
-      ContentRoot="$(YarnWorkingDir)dist\$(Configuration)\"
-      BasePath="_content/$(PackageId)"
-      CandidateAssets="@(_InteropBuildOutput)"
-      RelativePathFilter="**.js"
-    >
+    <DefineStaticWebAssets Condition="@(_InteropBuildOutput) != ''" SourceType="Computed" SourceId="$(PackageId)" ContentRoot="$(YarnWorkingDir)dist\$(Configuration)\" BasePath="_content/$(PackageId)" CandidateAssets="@(_InteropBuildOutput)" RelativePathFilter="**.js">
```
Comment (TanayParikh) [inline src/Components/WebAssembly/WebAssembly.Authentication/src/Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj:79]: Was this unwrapping intentional? If not, could we please revert this bit?

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Services/RemoteAuthenticationService.cs [hunk 3]
```diff
@@ -16,7 +16,7 @@ namespace Microsoft.AspNetCore.Components.WebAssembly.Authentication;
 /// <typeparam name="TRemoteAuthenticationState">The state to preserve across authentication operations.</typeparam>
 /// <typeparam name="TAccount">The type of the <see cref="RemoteUserAccount" />.</typeparam>
 /// <typeparam name="TProviderOptions">The options to be passed down to the underlying JavaScript library handling the authentication operations.</typeparam>
-public class RemoteAuthenticationService<TRemoteAuthenticationState, [DynamicallyAccessedMembers(JsonSerialized)] TAccount, TProviderOptions> :
+public class RemoteAuthenticationService<[DynamicallyAccessedMembers(JsonSerialized)] TRemoteAuthenticationState, [DynamicallyAccessedMembers(JsonSerialized)] TAccount, [DynamicallyAccessedMembers(JsonSerialized)] TProviderOptions> :
```
Comment (TanayParikh) [inline src/Components/WebAssembly/WebAssembly.Authentication/src/Services/RemoteAuthenticationService.cs:19]:
```suggestion
public class RemoteAuthenticationService<
[DynamicallyAccessedMembers(JsonSerialized)] TRemoteAuthenticationState,
[DynamicallyAccessedMembers(JsonSerialized)] TAccount,
[DynamicallyAccessedMembers(JsonSerialized)] TProviderOptions> :
```

File: src/Components/WebAssembly/WebAssembly.Authentication/src/WebAssemblyAuthenticationServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -26,7 +26,8 @@ public static class WebAssemblyAuthenticationServiceCollectionExtensions
     /// <typeparam name="TProviderOptions">The configuration options of the underlying provider being used for handling the authentication operations.</typeparam>
     /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
     /// <returns>The <see cref="IServiceCollection"/> where the services were registered.</returns>
-    public static IRemoteAuthenticationBuilder<TRemoteAuthenticationState, TAccount> AddRemoteAuthentication<TRemoteAuthenticationState, [DynamicallyAccessedMembers(JsonSerialized)] TAccount, TProviderOptions>(
+    public static IRemoteAuthenticationBuilder<TRemoteAuthenticationState, TAccount> AddRemoteAuthentication<
+        [DynamicallyAccessedMembers(JsonSerialized)] TRemoteAuthenticationState, [DynamicallyAccessedMembers(JsonSerialized)] TAccount, [DynamicallyAccessedMembers(JsonSerialized)] TProviderOptions>(
```
Comment (TanayParikh) [inline src/Components/WebAssembly/WebAssembly.Authentication/src/WebAssemblyAuthenticationServiceCollectionExtensions.cs:30]:
```suggestion
        [DynamicallyAccessedMembers(JsonSerialized)] TRemoteAuthenticationState,
        [DynamicallyAccessedMembers(JsonSerialized)] TAccount,
        [DynamicallyAccessedMembers(JsonSerialized)] TProviderOptions>(
```

**Review by TanayParikh — APPROVED** [review]
LGTM, just some formatting nits

**Comment by pranavkm** [comment]
Authentication.MSAL was already annotated for trimming. However WebAssembly.Authentication previously wasn't. Investigating this issue, the underlying issue was that our JSRuntime.Invoke* APIs aren't annotated accurately. Essentially, all of the `args` array is subject to serialization constraints, but we don't express this outside of slapping a `RequiresUnreferencedCode` on these methods. In this particular case, `MsalProviderOptions` is passed as a JSRuntime.InvokeAsync argument but was being trimmed (all of the getters were being removed). Annotating the entry point as this PR does helps, but it would help to revist JSRuntime's annotations.

**Comment by TanayParikh** [comment]
> Annotating the entry point as this PR does helps, but it would help to revist JSRuntime's annotations.

Do we have a tracking issue for this?

**Comment by SteveSandersonMS** [comment]
Great that you were able to resolve the issue with the authentication libraries!

> Investigating this issue, the underlying issue was that our JSRuntime.Invoke* APIs aren't annotated accurately. Essentially, all of the args array is subject to serialization constraints, but we don't express this. Unfortunately, the best we could do is slap a RequiresUnreferencedCode on these methods.

By that do you mean we're unable to annotate them because the right kinds of annotation don't exist and we'd need to ask for new compiler features, or do you mean we can and should put `[RequiresUnreferencedCode]` on the interface methods?

**Pull request merged**